#Trabalho Prático Análise de Dados:
#Exercitar os seguintes conceitos trabalhados no Módulo:
##✓ Criação e manipulação de numpy arrays.
##✓ Leitura de dados e manipulação de DataFrames e Series no pandas.
##✓ Fundamentos de aprendizado de máquinas.


In [ ]:
# importando as bibliotecas
import numpy as np
import pandas as pd

**Código 1**

In [ ]:
# == Código 1 ==
Z = np.zeros((4,))
print("Z:", Z)

Z: [0. 0. 0. 0.]


**Pergunta 1**: O que será “printado” na tela ao se executar o Código 1? *Z: [0. 0. 0. 0.]* 


**Código 2**

In [ ]:
# == Código 2 ==
Z = np.zeros((4,))
Z[1] = 1.
print("Z:", Z)

Z: [0. 1. 0. 0.]


**Pergunta 2**: O que será “printado” na tela ao se executar o Código 2? *Z: [0. 1. 0. 0.]* 

**Código 3**

In [ ]:
# == Código 3 ==
Z = np.zeros((4,))
Z[1:] = 1.
print("Z:", Z)

Z: [0. 1. 1. 1.]


**Pergunta 3**: O que será “printado” na tela ao se executar o Código 3? *Z: [0. 1. 1. 1.]* 

**Código 4**

In [ ]:
# == Código 4 a ==
Z = np.zeros((4,))
Z[:3] = 1.
print("Z:", Z)

Z: [1. 1. 1. 0.]


In [ ]:
# == Código 4 b ==
Z = np.ones((4,))
#Z[4] = 0.
Z[3] = 0.
print("Z:", Z)

Z: [1. 1. 1. 0.]


In [ ]:
# == Código 4 c ==
Z = np.zeros((4,))
Z[:-1] = 1.
print("Z:", Z)

Z: [1. 1. 1. 0.]


In [ ]:
# == Código 4 d ==
Z = np.ones((4,))
Z[-1] = 0.
print("Z:", Z)

Z: [1. 1. 1. 0.]


**Pergunta 4**: Qual dos seguintes trechos de código NÃO resulta no “print” mostrado no Código 4? *Z = np.ones((4,)) Z[4] = 0*

**Código 5**

In [ ]:
# == Código 5 a ==
#X = np.twos((2,2))
X = np.ones((2,2))
print("X:\n", X)

X:
 [[1. 1.]
 [1. 1.]]


In [ ]:
# == Código 5 b ==
X = np.array([2.]*4).reshape(2,2)
print("X:\n", X)

X:
 [[2. 2.]
 [2. 2.]]


In [ ]:
# == Código 5 c ==
X = 2 * np.ones((2,2))
print("X:\n", X)

X:
 [[2. 2.]
 [2. 2.]]


In [ ]:
# == Código 5 d ==
X = np.ones((2,2)) + np.ones((2,2))
print("X:\n", X)

X:
 [[2. 2.]
 [2. 2.]]


**Pergunta 5**: Qual dos seguintes trechos de código NÃO resulta no “print” mostrado no Código 5? *X = np.twos((2, 2))*

**Código 6**

In [ ]:
# == Código 6 ==
X = np.array([[1,2], [3, 4]])
Y = X[0, :]
Y[1] = 10
print("X:\n", X)
print("Y:\n", Y)

X:
 [[ 1 10]
 [ 3  4]]
Y:
 [ 1 10]


**Pergunta 6**: O que será “printado” na tela ao se executar o Código 6? *X: [[ 1 10] [ 3 4]]* 

**Código 7**

In [ ]:
# == Código 7 ==
X = np.array([[1,2], [11, 10]])
print(np.mean(X[X > np.pi]))
print("X:\n", X)
print(X > np.pi)
print(X[X > np.pi])

10.5
X:
 [[ 1  2]
 [11 10]]
[[False False]
 [ True  True]]
[11 10]


**Pergunta 7**: O que será “printado” na tela ao se executar o Código 7?  *A média dos valores de X que são maiores que np.pi (3.1415...).*

**Código 8**

In [ ]:
# == Código 8 ==
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
pd.DataFrame(data=data, index=labels)

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

**Pergunta 8**: Baseado no Código 8, o que o comando pd.DataFrame(data=data, index=labels) irá retornar? *Um pandas DataFrame baseado no dicionário data e indexado pela lista labels.*

**Código 9**

In [ ]:
# == Código 9 ==
df = pd.DataFrame(data=data, index=labels)
df.shape
#df.length
#df.rows_cols
#df.len

(10, 4)

**Pergunta 9**: Baseado nos Códigos 8 e 9, qual atributo de df pode ser acessado para uma rápida verificação de sua dimensão (número de linhas, número de colunas)? *df.shape*

In [ ]:
 c = df['animal'].value_counts()
 print(c)
 print('Series ou data frames \n', type(df['animal'].value_counts()))
 print('Series ou data frames \n', type(c))

cat      4
dog      4
snake    2
Name: animal, dtype: int64
Series ou data frames 
 <class 'pandas.core.series.Series'>
Series ou data frames 
 <class 'pandas.core.series.Series'>


**Pergunta 10**: Baseado nos Códigos 8 e 9, o que o comando df[‘animal’].value_counts() irá retornar? *Um pandas Series contendo a contagem da quantidade de vezes que cada elemento único aparece na coluna ‘animal’.* 

In [ ]:
df.describe()

age     visits
count  8.000000  10.000000
mean   3.437500   1.900000
std    2.007797   0.875595
min    0.500000   1.000000
25%    2.375000   1.000000
50%    3.000000   2.000000
75%    4.625000   2.750000
max    7.000000   3.000000

**Pergunta 11**: Baseado nos Códigos 8 e 9, qual a média (mean) da coluna ‘age’ que aparece ao se executar o comando df.describe()? *3.437500* 

In [ ]:
df.iloc[:, 3]
#df['visits']
#df.iloc[:, -2]
#df.loc[:, 'visits']

a    yes
b    yes
c     no
d    yes
e     no
f     no
g     no
h    yes
i     no
j     no
Name: priority, dtype: object

**Pergunta 12**: Baseado nos Códigos 8 e 9, qual dos comandos NÃO irá retornar um pandas Series contendo a coluna ‘visits’? *df.iloc[:, 3]*

In [ ]:
df.sort_values(by='visits', ascending=False)
#df.sort_index(ascending=False)


animal  age  visits priority
b    cat  3.0       3      yes
d    dog  NaN       3      yes
f    cat  2.0       3       no
c  snake  0.5       2       no
e    dog  5.0       2       no
i    dog  7.0       2       no
a    cat  2.5       1      yes
g  snake  4.5       1       no
h    cat  NaN       1      yes
j    dog  3.0       1       no

**Pergunta 13**: Baseado nos Códigos 8 e 9, qual dos seguintes comandos irá retornar um pandas DataFrame ordenado de forma decrescente baseado na coluna ‘visits’? *df.sort_values(by='visits', ascending=False)* 

**Código 10**

In [ ]:
# == Código 10 ==
y_true = np.array([1., 2., 1.])
y_pred = np.array([1.1, 1.98, 1.05])

#np.sqrt(((y_true-y_pred)**2).mean())
mse_reg = ((y_true - y_pred)**2).mean() #MSE (Mean Squared Error) = Erro Quadrático Médio
np.sqrt(mse_reg)

0.06557438524302006

**Pergunta 14**: Baseado no Código 10, seja y_true os valores que devem ser previstos a partir de um modelo de aprendizado de máquinas e seja y_pred os valores que de fato foram previstos por esse modelo. O que o comando np.sqrt(((y_true-y_pred)**2).mean()) irá retornar? *A raiz quadrada do erro quadrático médio do modelo.* 

In [ ]:
# função para cálculo do MSE
def mse(y_true, y_pred, is_ref = False):

  # mse modelo
  if is_ref:
    mse = ((y_true - y_true.mean())**2).mean()
  else:
    mse = ((y_true - y_pred)**2).mean()

  return mse

In [ ]:
# função para cálculo do coeficiente de determinação R2
def r2(mse_reg, mse_ref):
  return 1 - mse_reg/mse_ref

In [ ]:
# calculando o mse dos modelos
mse_reg = mse(y_true=y_true, y_pred=y_pred)
print("MSE do modelo de regressão:", mse_reg)
mse_ref = mse(y_true=y_true, y_pred=y_pred, is_ref=True)
print("MSE do modelo de referência:", mse_ref)

MSE do modelo de regressão: 0.004300000000000008
MSE do modelo de referência: 0.2222222222222222


In [ ]:
# calculando o R2 score
r2_score = r2(mse_reg=mse_reg, mse_ref=mse_ref)
print("Coeficiente R2 do modelo implementado (calculado):", r2_score)

Coeficiente R2 do modelo implementado (calculado): 0.9806499999999999


In [ ]:
((y_true-y_pred)**2).mean()

0.004300000000000008

In [ ]:
# == Código 10 ==
y_true = np.array([1., 2., 1.])
#y_pred = np.array([1.1, 1.98, 1.05])
#y_pred = np.array([0, 0, 0])
#@y_pred = np.array([1, 1, 1])
#y_pred = np.array([-1., -2., -1.])
y_pred = np.array([1., 2., 1.])

#np.sqrt(((y_true-y_pred)**2).mean())
#MSE (Mean Squared Error) = Erro Quadrático Médio
((y_true-y_pred)**2).mean()

0.0

**Pergunta 15**: Baseado no Código 10, qual o valor de y_pred para que o erro quadrático médio de predição seja nulo? *y_pred deve ser igual a y_true.* 